pkg/operator/provisioner.go

```diff
@@ -0,0 +1,207 @@
+
+	capacity := options.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]
+	requestBytes := capacity.Value()
+
+	imageName := fmt.Sprintf("kubernetes-dynamic-%s-%s", options.PVName, uuid.NewUUID())
+
+	rookClient, err := p.clusterManager.getRookClient(p.provConfig.clusterNamespace)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get rook client: %v", err)
+	}
+
+	res, err := createVolume(imageName, p.provConfig.pool, requestBytes, rookClient)
+	if err != nil {
+		return nil, err
+	}
+	glog.V(1).Infof("Rook block image created: %s", res)
+
+	rookClientInfo, err := rookClient.GetClientAccessInfo()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get rook client information: %v", err)
+	}
+	monitors := processMonAddresses(rookClientInfo.MonAddresses)
+	radosUser := fmt.Sprintf("%s-rbd-user", p.provConfig.clusterName)
+	secretRef := new(v1.LocalObjectReference)
+	secretRef.Name = fmt.Sprintf("%s-rbd-user", p.provConfig.clusterName)
+
+	pv := &v1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: options.PVName,
+		},
+		Spec: v1.PersistentVolumeSpec{
+			PersistentVolumeReclaimPolicy: options.PersistentVolumeReclaimPolicy,
+			AccessModes:                   options.PVC.Spec.AccessModes,
+			Capacity: v1.ResourceList{
+				v1.ResourceName(v1.ResourceStorage): capacity,
+			},
+			PersistentVolumeSource: v1.PersistentVolumeSource{
+				RBD: &v1.RBDVolumeSource{
+					RBDImage:     imageName,
+					RBDPool:      p.provConfig.pool,
+					CephMonitors: monitors,
+					RadosUser:    radosUser,
+					SecretRef:    secretRef,
+					FSType:       "ext4",
+					ReadOnly:     false,
+				},
+			},
+		},
+	}
+	glog.Infof("successfully created Rook Block volume %+v", pv.Spec.PersistentVolumeSource.RBD)
+	return pv, nil
+}
+
+// createVolume creates a rook block volume.
+func createVolume(image, pool string, size int64, client rookclient.RookRestClient) (string, error) {
+	newImage := model.BlockImage{
+		Name:     image,
+		PoolName: pool,
+		Size:     uint64(size),
+	}
+
+	res, err := client.CreateBlockImage(newImage)
+	if err != nil {
+		return "", fmt.Errorf("Failed to create rook block image %s/%s: %v", pool, image, err)
+	}
+
+	return res, nil
+}
+
+// Delete removes the storage asset that was created by Provision represented
+// by the given PV.
+func (p *rookVolumeProvisioner) Delete(volume *v1.PersistentVolume) error {
+	rookClient, err := p.clusterManager.getRookClient(p.provConfig.clusterNamespace)
+	if err != nil {
+		return fmt.Errorf("Failed to get rook client: %v", err)
+	}
+
+	image := model.BlockImage{
+		Name:     volume.Spec.PersistentVolumeSource.RBD.RBDImage,
+		PoolName: p.provConfig.pool,
+	}
+
+	_, err = rookClient.DeleteBlockImage(image)
+	if err != nil {
+		return fmt.Errorf("Failed to delete rook block image %s: %v", volume.Name, err)
+	}
+	return nil
+}
+
+func parseClassParameters(params map[string]string) (*provisionerConfig, error) {
+	var cfg provisionerConfig
+	defaultCluster := "rook"
```

Comment:
I think it'd be nice to have this in a const somewhere perhaps in the cluster or operator pkg.